### ISSUE TITLE
NUnit 3.0.1 AssemblyVersion is diffrent for .NET 4.0 and .net 4.5

### ISSUE THREAD

**Issue by shatl** [nunit/nunit#1108]
.NET 4.5 assembly has AssemblyVersion 3.0.5813.39031
.NET 4.0 assmebly has AssemblyVersion 3.0.5813.39032

**Comment by rprouse** [comment]
Does this cause a problem? The assemblies are currently versioned using 3.0.*.*, so if they are compiled at slightly different times they will end up with different numbers. These were compiled at the same time, but I guess 4.0 started a bit after 4.5?

**Comment by CharliePoole** [comment]
I've noticed this as well. It bothers me because it makes me stop and think "What does this mean?" but I don't think it has any more effect than that.

In any case, when we update our versioning for 3.1, we may want to get rid of the stars entirely for production builds and force in our own coded value.

Let's keep this as a build issue, in order to track the problem.

**Comment by shatl** [comment]
Will be a compiler error if .NET4.5 assembly references NUnit and another .NET4 assembly which references .NET4.

**Comment by CharliePoole** [comment]
So is this what you are saying?

```
    test.assembly (4.5) => nunit.framework (4.5)
                                   => someAssembly (4.0) => nunit.framework (4.0)
```

I'm not sure why you would do this, but it wouldn't work anyway, since you would be trying to load two different nunit.framework assemblies (different identities) into the same AppDomain.

Or do you mean something else?

**Comment by corliss** [comment]
This is causing my project not to load. The fusion log shows that it's trying to load `3.0.5813.39035` (the version in lib/dotnet) but found `3.0.5813.39031` (the version in lib/net45), resulting in a "The assembly reference did not match the assembly definition found" error.

This occurs in a UI test application that is loading assemblies via reflection. These assemblies have a combination of UI tests and unit tests. The unit test code (or nunit) is not the reason these assemblies are being loaded. However, this load error is preventing the app from executing.

Under what circumstances would the `lib/dotnet` version of the code load vs the `lib/net45` version?

**Comment by CharliePoole** [comment]
Short answer: It will always load the one you are referencing.

For a given assembly, you can only be referencing one of them. If you are using the nuget package, the reference is created automatically but you can examine and change it if necessary.

If your UI test application references the nunit framework, it can never load an assembly that uses a different build of the framework into the same appdomain. You can, of course, create a separate appdomain in which to load it.

Even if your test app does not reference the framework, it may be trying to load different assembiles with references to different versions of the framework into the same appdomain. That won't work either.

The conflicting version numbers are accidental and we plan to remove fix that in a coming release. That would not, however, fix your problem. Each assembly needs the framework for which it was compiled.

One long term solution might be to avoid mixing unit tests with UI tests in the same assembly.

**Comment by corliss** [comment]
Thanks, but the term "you" is confusing.

There are multiple assemblies here, each separately referencing the NUnit nuget package. They don't reference an NUnit assembly directly, only via the package.

So, given the scenario above, when does the "lib/dotnet" flavor be used, and when does the "lib/net45" flavor be used? Can you explain the behavior described above, and what I can do about it.

There are no appdomains involved here. This is a fairly simple scenario.

Thanks!

**Comment by CharliePoole** [comment]
I'm assuming you are the person who created this solution. If not, you may silently substitute a different pronoun.

When you reference the nuget package, nuget adds an assembly reference. You should expect it to be the correct one. I suggest you check.

The dot net flavor should be used for a PCL. The 4.5 assembly should be used for an assembly that targets 4.5. Even though nuget does it for you, you should make sure it's right. Probably it is.

If you are loading multiple assemblies with potentially conflicting references then the fact that "there are no appdomains involved" is the source of your problem.

OTOH if the conflicts are unintended and can be removed, you can continue without creating any domains.

**Comment by corliss** [comment]
(Yes, I've created the solution and all the assemblies, so the metonym "you" was not clear - which assembly did you mean. Anyway enough of that :) )

A more detailed description:

a) The executable (A) is a .net 4.5 binary. Since it references the NUnit nuget package, the lib/net45 flavored assembly is loaded.

b) The executable loads a PCL (B) that also references the NUnit nuget package. No errors occur at this step. Based on what you said, it ought to load the "lib/dotnet" flavor. Does fusion silently ignore the assembly mismatch here?

c) The same library B introspects on a type, and loads its Assembly using `type.GetTypeInfo().GetAssembly()`. Interestingly, the assembly happens to be the same as the currently executing assembly, B.

d) At this point the failure occurs.

So let's just look at the intent here. A .net 4.5 executable implicitly loads a PCL. This PCL loads its own Assembly via reflection. Both the executable and the PCL reference the same NUnit nuget package.

Why should this *scenario* fail? It looks like something in the way PCL's or Nuget work together isn't quite right.

Am I missing something

**Comment by rprouse** [comment]
@corliss, the behavior you are seeing what I would expect for your setup. It would happen with any NuGet package that you reference from A and B that provides .NET 4.5 and PCL versions. Even though they are named the same, signed the same and provide the same functionality, a PCL and a .NET 4.5 assembly are different assemblies. Even if the version numbers were the same, it still wouldn't work.

Since Assembly B references NUnit, I assume it is a test assembly? Why is it a PCL? If you make it target 4.5, everything should work.

What platform are you trying to create tests for?

**Comment by corliss** [comment]
No, it isn't a test assembly. It is a UI test harness. That's why it is introspecting on other assemblies. Those test assemblies are also PCLs. Only the loader knows about the platform, and all platform specific services are injected.

The loader story with NuGet packages is starting to get really ugly and complex, I've encountered other strange bugs. Not your fault, of course.

For NUnit specifically, why are different assemblies provided for .net and PCL? I'd expect the same binary in both the lib/dotnet and lib/net45 directories, since the PCL version is a subset of the latter.

Is that reasonable? In which case I think it is a genuine NUnit bug (albeit not your "fault") that can be fixed by providing the same binary with the same revision in both lib/dotnet and lib/net45.

**Comment by rprouse** [comment]
@corliss, you stated it yourself, the PCL is a subset of the .NET 4.5. Because it is a subset, it does not provide all the functionality that the 4.5 version does. For example, it is missing most of the file and directory asserts. It also does not contain the serialization or .NET remoting code that allows test results to be sent across process boundaries.

It sounds like you are writing your own test runner. If assembly A is a test runner that executes tests through assembly B, then you should reference the portable assembly from Assembly A. It is intended that if you use the portable framework anywhere, then you use it everywhere. What you are doing sounds similar to what we do in the Xamarin runners, but unfortunately the .NET 4.5 framework is getting in your way.

**Comment by CharliePoole** [comment]
If B is not a unit test, there should be no need for it to reference NUnit at all. 

The PCL version of the framework is used for things like Windows 8.1 and Xamarin.IoS. It is also capable of being loaded by .NET 4.5 tests, but isn't the best choice since it lacks some features that are in the .NET 4.5 build of the framework.

The version numbering will be fixed when we next increment the minor component. However, as @rprouse indicates, any assembly conflict would still exist. It would just give you a more subtle error.

**Comment by corliss** [comment]
Thanks for the replies. However, there's still a loose end: why does the assembly load and execute in step b but fail in step c, reflecting on itself?

I'd like to ask also, how confident are you of whatever explanation you provide for this inconsistency? Because if none of us can completely explain it, perhaps it needs to be addressed by the .net/fusion folks. So please do give an answer but also your own confidence in it.

**Comment by CharliePoole** [comment]
My guess is that assembly B's code is not actually attempting to reference the framework in any way and that if it did, you would get the error sooner. An experiment would be needed to give me confidence in this answer.

It could also have to do with how you are loading the assembly in each case. Does A simply reference B or is it loaded dynamically in some way?

I have run into similar problems with reflection myself, since NUnit relies heavily on reflection. My experience is that a separate AppDomain is required in all practical cases. That is, one can start out without it and see that things appear to work but eventually a crisis arises that requires the AppDomain. This is merely anecdotal experience of course.

It's extremely likely that we understand this less well than we should. :-(

**Comment by corliss** [comment]
@CharliePoole A simply references B and calls into it. B reflects on itself via `type.GetTypeInfo().GetAssembly()`.

**Comment by rprouse** [comment]
@corliss can you edit the project file and change the HintPath to the portable library?

As for level of confidence, very high. I would never expect what you are doing to work, the NUnit PCL and .NET 4.5 are different, you cannot expect them to behave the same.

It likely crashes when you are reflecting through it because that is the first time that you access NUnit. I assume that in your code in B, you do not call any NUnit code before you do the reflection. .NET does not load referenced assemblies until they are called, not on startup, but depending on how you are doing the reflection, it will cause dependent assemblies to be loaded.

**Comment by CharliePoole** [comment]
@corliss Then as @rprouse says, it's simply that you are not actually accessing the code.

I'll add to Rob's recommendations: remove the reference entirely where it is not needed. When you reference a package, NuGet assumes all your projects need to refer to it. It's up to you to clear the check boxes for the ones that don't actually use it - in this case the non-unit-test assemblies.

**Comment by corliss** [comment]
@CharliePoole: As I've explained already, the reference *is* needed, so please let's focus on the cause, not on recipes. Agreed? I just need to understand the cause well enough.

As an aside, here's an interesting cracker. You're telling me to remove the reference to the NUnit nuget from A. The exact opposite was needed to work around this [issue](https://github.com/ericsink/SQLitePCL.raw/issues/61) with the SQLitePCL.raw nuget package. That package uses the "bait and switch" technique to load the right assembly, so you actually *have to* reference the assembly from the executable even though the compiler does not need it.

It both cases, there does not seem much that package authors can do about it - the .net/nuget/PCL loaders decide what to load.

@rprouse, Changing the `HintPath` to the `dotnet` version does work. This is pretty fragile, obviously, and will break with the next Nuget update. Is this the best we can do?

**Comment by rprouse** [comment]
@corliss unfortunately, if you want NuGet to manage your packages, you have to work within its limitations and accept that it is fragile. Sorry, I can't think of a better alternative.

**Comment by CharliePoole** [comment]
Unfortunately, __we__ have to understand what you are doing in order to help you. That means we may ask questions that you consider unfocused or useless.

Whatever reason you have for needing the reference, you have to make all the references point to the same build of the framework or one of them will fail to load. You can do this by removing references, re-pointing them to the proper lib directory or changing the hint path.

If your issue is that you don't think you should **have to** do this, I couldn't agree more. However, it's out of our control. We are users of Microsoft .NET and of NuGet, just as you are.

'nuff on this for me!

**Comment by CharliePoole** [comment]
@rprouse In spite of this digression, I guess the original intent of this issue - to ensure consistent version numbers in 3.2 - still stands. Yes?

**Comment by rprouse** [comment]
Yes, I think that consistent versions are desirable.

**Comment by corliss** [comment]
@CharliePoole thanks for your input. It has nothing to do with what I want or how I think it should be, it is just about understanding the system, warts and all, and then doing the best.

@rprouse, thanks as well - are you comfortable with the recommendation? Have you recommended this to others?  To be sure, I haven't felt that Nuget was fragile until I ran into both these issues (this one and the one I referenced in SQLitePCL.raw), both of which surfaced for me only recently. 

I do think that this issue is likely to be hit by many people. Consistent versioning will mask it to where it becomes an issue only for those referencing platform-specific functions. In other words, a really bad leaky nuget abstraction. Again **not** your fault, but something your users may well hit, and probably worth warning about.

**Comment by rprouse** [comment]
I am not sure how often it will hit other people. I do a lot of testing on various platforms using a mix of PCL and native versions of the framework and it has always just worked. I think your usage is an edge case, but I guess time will tell.

**Comment by corliss** [comment]
OK, thanks. If you think the consistent version fix can happen soon, that will save a bunch of work here and will be much appreciated.

**Comment by rprouse** [comment]
@corliss, unless we do a 3.0.2 release, the consistent versioning will
likely be late January'ish with the 3.2 release. I also don't think it will
fix your issues as you described them.

**Comment by corliss** [comment]
@rprouse Could you elaborate? I thought I understood a bit, but your reply tells me I haven't. Why wouldn't consistent numbering fix, or more properly, mask my issue? Right now the loader fails because of an incorrect revision, not because of code differences between the assemblies.

**Comment by rprouse** [comment]
@corliss consistent versioning may get you past the crash, but as you say,
it will only mask the issue. It will likely crash later or give you odd
bugs. You might be okay since the PCL is a subset of the .NET 4.5 version,
but I can't guarantee it.

BTW, are you creating your own runner, or will you end up using one of the
NUnit runners?

**Comment by corliss** [comment]
> @corliss consistent versioning may get you past the crash, but as you say,
> it will only mask the issue. It will likely crash later or give you odd
> bugs. You might be okay since the PCL is a subset of the .NET 4.5 version,
> but I can't guarantee it.

The point is that the mask is the fix in this case. Will take your answer as a low-confidence one. Can't say I understand the superstitious language - at this point, having drilled into the root cause am personally pretty confident that there will *not* be such issues. And the test, using the temporary `HintPath` fix, bears this out.

[The fact that NUnit has differently-able assemblies for the same package is cause for concern, but no more so for my package than for the other mix of platforms you say you've tested. What's more disconcerting is that the nuget/pcl mechanism leaves package owners unable to make confident statements about use cases. My use case is utterly ordinary and yet has resulted in this thread.]

So I look forward to consistent version numbers, and do recommend a hot fix. It may save a bloke or two having to go through the same effort. (This took a few days of denial & then investigation before I came here.)

**Comment by corliss** [comment]
Hi - I keep hitting this issue. Would it be possible to please fix the versions of the binaries to be consistent?

The problem has been described in detail earlier in this thread. This is not a theoretical issue, and all that's needed to fix this is to have the same version on the various binaries. Pretty please. Puh-leeze.

Thanks!

**Comment by CharliePoole** [comment]
@corliss I haven't been following this discussion for a while so I'm confused. What do you mean by "newer releases of NUnit?" You filed this issue on NUnit 3.0.1 and there have been no further releases.

**Comment by corliss** [comment]
You're right, my bad - I was sleepy last night when I wrote this. But I do keep hitting this. Could you give an ETA? Thanks!

**Comment by CharliePoole** [comment]
This was actually fixed with the 3.2 release.
